contrib/experimental/great_expectations_experimental/expectations/expect_column_pair_correlation_to_be_between.py

```diff
@@ -0,0 +1,385 @@
+import json
+from typing import Any, Dict, Optional, Tuple
+
+import numpy as np
+import pandas as pd
+from scipy import stats as stats
+
+from great_expectations.core import ExpectationConfiguration
+from great_expectations.execution_engine import (
+    ExecutionEngine,
+    PandasExecutionEngine,
+    SparkDFExecutionEngine,
+)
+from great_expectations.execution_engine.execution_engine import MetricDomainTypes
+from great_expectations.execution_engine.sqlalchemy_execution_engine import (
+    SqlAlchemyExecutionEngine,
+)
+from great_expectations.expectations.expectation import (
+    ColumnExpectation,
+    ColumnPairExpectation,
+    Expectation,
+    ExpectationConfiguration,
+    InvalidExpectationConfigurationError,
+    _format_map_output,
+)
+from great_expectations.expectations.metrics.column_aggregate_metric_provider import (
+    ColumnMetricProvider,
+    ColumnPairAggregateMetricProvider,
+    column_aggregate_value,
+    column_pair_aggregate_value,
+)
+from great_expectations.expectations.metrics.import_manager import Bucketizer, F, sa
+from great_expectations.expectations.metrics.metric_provider import (
+    MetricProvider,
+    metric_value,
+)
+from great_expectations.expectations.util import render_evaluation_parameter_string
+from great_expectations.render.renderer.renderer import renderer
+from great_expectations.render.types import RenderedStringTemplateContent
+from great_expectations.render.util import (
+    handle_strict_min_max,
+    num_to_str,
+    parse_row_condition_string_pandas_engine,
+    substitute_none_for_missing,
+)
+from great_expectations.validator.validation_graph import MetricConfiguration
+
+
+class ColumnCorrelation(ColumnPairAggregateMetricProvider):
+    """MetricProvider Class for column pair correlation"""
+
+    metric_name = "column_pair.correlation"
```

Comment:
@edjoesu This is where I feel that we have an opportunity to create the class variable name for Aggregate Metrics that is distinct from and is a sibling of the ones for Map Metrics and Condition Metrics.  Then this would tie in nicely with the new `Expectation` class you created (`ColumnPairAggregateExpectation`) and the underlying new Metrics class `ColumnPairAggregateMetricProvider` (along with the decorators).

On a separate note, I can see the need to remove new formatting added to `expectation.py` and also clean up the code for `column_filtering` in `column_aggregate_metric_provider.py`.  We would also need to upgrade `column_aggregate_metric_provider.py` in general (potentially to include `ignore_row_if`), and the way it uses `get_compute_domain()` may not be according to the latest recommended coding pattern.  Thanks!

Reply:
@alexsherstinsky it took me a while to parse this, but I think I've wrapped my head around the relevant issues flagged in this review. Please check/confirm/revise:

(1) "create the class variable name for Aggregate Metrics that is distinct from and is a sibling of the ones for Map Metrics and Condition Metrics" --> I think that `column_pair` fits that bill. In the single column case, we have `column.<name>` for aggregates, and `column_values.<name>` for maps; in this case, we have `column_pair.<name>` for aggregates and `column_pair_values.<name>` for maps.

(2) I did revise the PR to accept your above-proposed refactor and renamed "ColumnPairExpectation" to "ColumnPairAggregateExpectation"

(3) re: the refactoring and cleanup -- strong agree, but I think we can move it out of the scope of this PR (since it involves other expectations as well).

Reply:
@jcampbell Thanks to the note by @abegong I finally saw your comment.  Sorry that I missed it before.

In my opinion, a solution that would enable the kind of user (developer) experience for `column_pair` and `multi-column` aggregate metrics and expectations that are currently present for single-column aggregate metrics and expectations would need to include the following items:
* Develop decorators that are analogous to `column_aggregate_value` and `column_aggregate_partial` for the column-pair and multi-column use cases.
* Convert the existing aggregate metric implementation classes for column-pair and multi-column domains, which are currently using the lowest level metric decorators, to take advantage of these higher level decorators, and thereby delete repetitive common code from these classes, thereby leveraging the benefits of the GX framework.
* Aggregates can be bundled (we utilize complex logic for bundling single-column aggregate metric computations); hence, we should investigate whether or not bundling can also be accomplished for column-pair aggregate metrics and multi-column aggregate metrics.  If affirmative, then implement it; otherwise, develop a different way of running “execute()“/”collect()” for `SQLAlchemy` and `Spark`, respectively, for these new aggregate metrics.
* Develop the `ColumnPairAggregateExpectation` and `MulticolumnAggregateExpectation` base classes in "great_expectations/expectations/expectation.py" with common methods for obtaining computed metrics and reporting the expectation validation results, which inherited custom Expectation implementations can use out of the box (analogously to the way this is currently working for the Map Expectation implementations).
* Consider any implications to supporting `ignore_row_if` conditions.
* Write unit tests for the above new methods and classes.

With the above comprehensive list of capabilities in place, users will be equipped with sufficient framework support to develop custom column-pair aggregate metrics and multi-column aggregate metrics with no more difficulty than they can do today for single-column aggregate metrics and the corresponding custom expectations.

Thank you.